Change attributes navigation flow

ThomazFB opened the pull request:
Summary
==========

Fixes issue #4011 and #4013 by making available an overhaul at the Attributes view group flow following the new proposed UI

Also fixes #4011. 

| Variation Attributes now available at the Product Details View as a clickable card | Add Attribute view with a `next` button that only shows up when there are valid attributes selected | Add Attribute terms view with a `next` button that only shows up when there are at least one term selected | Attribute List view with a `done` button that only shows up when there are at least one attribute configured | Variation List no longer have access to the Attribute editing flow as a menu action |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| ![screenshot-1622648193185](https://user-images.githubusercontent.com/5920403/120590799-82084400-c411-11eb-9ba3-5a4f3cf00193.png) | ![screenshot-1622648070763](https://user-images.githubusercontent.com/5920403/120590805-85033480-c411-11eb-8ce5-0b366bd8ddac.png) | ![screenshot-1622648087642](https://user-images.githubusercontent.com/5920403/120590815-86ccf800-c411-11eb-9863-1eaf497789e4.png) | ![screenshot-1622648094787](https://user-images.githubusercontent.com/5920403/120590827-89c7e880-c411-11eb-89a5-27e5aabe07ab.png) | ![screenshot-1622648211072](https://user-images.githubusercontent.com/5920403/120590835-8af91580-c411-11eb-9e26-fff6529a1b4d.png) |

How to Test
==========
## Flow 1 - Generating a new Variation
https://user-images.githubusercontent.com/5920403/120590240-9ac42a00-c410-11eb-9f77-d2817b92b028.mp4

## Flow 2 - Modifying Attributes when generating a new variation 
https://user-images.githubusercontent.com/5920403/120590312-baf3e900-c410-11eb-944b-e357d49aed84.mp4

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Change done button to next at Attributes management views

ThomazFB committed:
Define isVariationCreation on AddAttributeTermsFragment

ThomazFB committed:
Add navigation action between AddAttributeTermsFragment and AttributeList

ThomazFB committed:
Create next action item at menu_attribute_terms

ThomazFB committed:
Handle isVariationCreation flow under AddAttributeTermsFragment

ThomazFB committed:
Connect variation creation flow flag between attribute fragments

ThomazFB committed:
Make AttributeListAdapter onItemClick an inline property

ThomazFB committed:
Merge branch 'develop' into issue/4011-change-attributes-navigation-flow

ThomazFB committed:
Fix navigation order when creating attributes

ThomazFB committed:
Set AttributeListFragment action button as done

ThomazFB committed:
Make next button visible only when changes are made to any the current Attribute

ThomazFB committed:
Change next button visibility to be visible when there's selected terms

ThomazFB committed:
Fix navigation order between attribute creation views

ThomazFB committed:
Make next button visible only when there's valid attributes

ThomazFB committed:
Merge branch 'develop' into issue/4011-change-attributes-navigation-flow

ThomazFB committed:
Save changes when moving to next attribute view

ThomazFB committed:
Define Variations attributes button at Product Details

ThomazFB committed:
Remove AttributeList reference from Variation List View

ThomazFB committed:
Create route between Product Detail view and Attribute List View

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Rename VariationListRepository to VariationRepository

ThomazFB committed:
Make ProductDetailViewModel capable of creating new variations

ThomazFB committed:
Change variation creation dialog text

ThomazFB committed:
Create route between AttributeListFragment and ProductDetailFragment

ThomazFB committed:
Start variation creation when AttributeListFragment done button gets hit

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Remove variation creation flow from Variation List View

ThomazFB committed:
Fix tests

ThomazFB committed:
Remove numVariation as a trackable change from a Product

ThomazFB committed:
Make AttributeListFragment done button only available when there's attributes

ThomazFB committed:
Change how Add Attribute button works during variation creation at Attribute list

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt:
```diff
@@ -103,7 +103,6 @@ data class Product(
             sku == product.sku &&
             slug == product.slug &&
             type == product.type &&
-            numVariations == product.numVariations &&
```
The `numVariations` field is a count controlled by the site, and it's modified as Variations get deleted or created, this doesn't essentially change the Product entity, since Product and Variations are treated as separate models connected through remote ID mapping, using this field as a modification condition causes the UI to think that the Product was changed and requires an update, when actually if this field changed the update already happened at the site.

With that said, I'm removing this field from the `isSameProduct` method to remove the bug where the `update` button show up at the `ProductDetailFragment` when it shouldn't.

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListFragment.kt:
```diff
@@ -111,30 +98,6 @@ class VariationListFragment : BaseDaggerFragment(R.layout.fragment_variation_lis
         }
     }
 
-    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
-        super.onCreateOptionsMenu(menu, inflater)
-
-        val mnuEditAttr = menu.add(Menu.NONE, ID_EDIT_ATTRIBUTES, Menu.NONE, R.string.product_variations_edit_attr)
-        mnuEditAttr.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
-        mnuEditAttr.isVisible = false
-    }
-
-    override fun onPrepareOptionsMenu(menu: Menu) {
-        super.onPrepareOptionsMenu(menu)
-
-        menu.findItem(ID_EDIT_ATTRIBUTES)?.isVisible = !viewModel.isEmpty
-    }
-
-    override fun onOptionsItemSelected(item: MenuItem): Boolean {
-        return when (item.itemId) {
-            ID_EDIT_ATTRIBUTES -> {
-                viewModel.onAddEditAttributesClick()
-                true
-            }
-            else -> super.onOptionsItemSelected(item)
-        }
-    }
-
```
Since the `VariationListFragment` no longer have access to the Attributes view flow, I'm removing all code related to the menu item calling the `AddAttributeFragment` navigation route and moving this call to the ProductDetailFragment through the ProductDetailCardBuilder

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt:
```diff
@@ -539,6 +542,26 @@ class ProductDetailCardBuilder(
             }
         )
 
+    private fun Product.variationAttributes() =
+        takeIf { this.numVariations > 0 && this.variationEnabledAttributes.isNotEmpty() }?.let {
+            val properties = mutableMapOf<String, String>()
+            for (attribute in this.variationEnabledAttributes) {
+                properties[attribute.name] = attribute.terms.size.toString()
+            }
+
+            PropertyGroup(
+                string.variable_product_attributes,
+                properties,
+                drawable.ic_gridicons_customize,
+                propertyFormat = string.product_variation_options
+            ) {
+                viewModel.onEditProductCardClicked(
+                    ViewProductAttributes,
+                    PRODUCT_ATTRIBUTE_EDIT_BUTTON_TAPPED
+                )
+            }
+        }
+
```
Relates to this: https://github.com/woocommerce/woocommerce-android/pull/4102/files#r644477680

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationRepository.kt:
```diff
@@ -28,7 +28,7 @@ import org.wordpress.android.fluxc.store.WCProductStore.OnProductChanged
 import javax.inject.Inject
 
-class VariationListRepository @Inject constructor(
+class VariationRepository @Inject constructor(
```
This repository owns behaviors related to Variations that must be available at other Variation-related viewModels, so I'm renaming it to a more wide Variation responsibility instead of being a repository made only for the `VariationListViewModel`

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Update release notes

ThomazFB committed:
Merge branch 'develop' into issue/4011-change-attributes-navigation-flow

ThomazFB commented:
Hey @Garance91540, this PR delivers your last change proposal around Variations, if possible, I would like to hear from you if this meets the intended objective.

Additionally, this PR also implements the changes around the `next` button we discussed in other threads. 

The UI issues you pointed up at the Attribute views are getting delivered here: https://github.com/woocommerce/woocommerce-android/pull/4125

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt:
```diff
@@ -501,16 +504,16 @@ class ProductDetailCardBuilder(
-            for (attribute in this.variationEnabledAttributes) {
-                properties[attribute.name] = attribute.terms.size.toString()
-            }
+            val content =
+                if (numVariations > 1)
+                    resources.getString(string.product_variation_multiple_count, numVariations.toString())
+                else
+                    resources.getString(string.product_variation_single_count)
```
np: We have a [method](https://github.com/woocommerce/woocommerce-android/blob/f58a86ae0a17c44a6102f6770c7032d53e539dbd/WooCommerce/src/main/kotlin/com/woocommerce/android/util/StringUtils.kt#L71) in `StringUtils` that simplifies this if else condition. Example [here](https://github.com/woocommerce/woocommerce-android/blob/b86665115ee9066966a716e3952f860af2ee9622/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt#L183-L182). What do you think about using that?

ThomazFB replied:
Hey Anitaa! Awesome suggestion, I wasn't aware of this function, good tip! I've added it here: 8ae10d8ff363398ddd626cb799595745f0130e01

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:
```diff
@@ -330,6 +335,21 @@ class ProductDetailViewModel @AssistedInject constructor(
+        attributeListViewState = attributeListViewState.copy(isCreatingVariationDialogShown = true)
+        launch {
+            viewState.productDraft?.let { draft ->
+                variationRepository.createEmptyVariation(draft)
+                    ?.let { updateProductDraft(numVariation = draft.numVariations + 1) }
+                    ?.let { triggerEvent(ExitProductAttributeList(variationCreated = true)) }
+                    ?: triggerEvent(ExitProductAttributeList())
+            }.also {
+                attributeListViewState = attributeListViewState.copy(isCreatingVariationDialogShown = false)
+            }
```
Nice <3

anitaa1990 reviewed (commented):
The code changes look good @ThomazFB! Nice work 🥇 I just left a few nitpicks in the code.

And had a few minor issues while testing:
 - The Next button is visible in Add Attribute and Add Attribute terms screen even when I haven't changed anything. Is that expected? The PR description says the button will only be displayed if there have been some changes to the screen.
<img src="https://user-images.githubusercontent.com/22608780/120626239-190be500-c480-11eb-966b-c30369357cb2.png" width="300"/>


- The click area of the `X` icon could be a bit bigger. From what I see in other image icons, we seem to be setting a fixed width and height of `image_major_50`. I'm not sure if that is the correct process to follow but it would be nice to increase that click area.
<img src="https://user-images.githubusercontent.com/22608780/120626193-0b565f80-c480-11eb-95de-643757cdb758.png" width="300"/>


- If the attribute terms are not clickable, it would be nice to remove the selectable background here. It feels misleading.
<img src="https://user-images.githubusercontent.com/22608780/120626056-ecf06400-c47f-11eb-9b64-b86c016b5832.png" width="300"/>

ThomazFB committed:
Disable selectableItemBackground when the term is already selected

ThomazFB committed:
Improve remove term button ripple effect

ThomazFB committed:
Improve variation card button value composition

ThomazFB committed:
Fix lint issues

ThomazFB commented:
Thank you so much for taking some of your time to review this PR, @anitaa1990. The PR is ready for a second round and I've submitted changes regarding your comments:

> The Next button is visible in Add Attribute and Add Attribute terms screen even when I haven't changed anything. Is that expected? The PR description says the button will only be displayed if there have been some changes to the screen.

Yes, this is expected! I did that intentionally. I decided to set the `Next` button visibility rule to consider if there are at least one Attribute properly configured to proceed with a Variation creation, I've done so because a Product can have the Attributes configured beforehand if it's not a new Variable Product scenario, so instead of forcing the merchant to do changes that aren't necessarily needed, they can just create the First variation with the current Product Attributes configuration.

In a scenario where the user is creating the Variable Product from scratch, since there will be no Attributes configured previously, then the `Next` button will appear only if there's any change done there and the scenario you described will be fulfilled. This behavior is also true for the `Done` button at `AttributeListFragment`, meanwhile, the `AddAttributeTermsFragment` `Next` button expects the selected Terms list to include at least one Term.

> The click area of the X icon could be a bit bigger. From what I see in other image icons, we seem to be setting a fixed width and height of image_major_50. I'm not sure if that is the correct process to follow but it would be nice to increase that click area.

I agree! I decided to use the borderless selection background to enable the click ripple to be a bit bigger and circular to follow the Android common behavior for this kind of icon button, let me know how this goes, here are the changes: acf330d6580bcc291dc35ffeec601ef0e289ce1f

>If the attribute terms are not clickable, it would be nice to remove the selectable background here. It feels misleading.

Good catch! I've fixed this here: d923d729067a79d3700a5bb01e929c8f14993c5c

Let me know what you think about these changes and thank you!

Garance91540 reviewed (changes requested):
Hi Thomaz, nice work, here are my notes:

<img width="551" alt="Screen Shot 2021-06-04 at 8 56 42 AM" src="https://user-images.githubusercontent.com/28450084/120759003-cc251e80-c512-11eb-8921-8deed6cbd525.png">

<img width="638" alt="Screen Shot 2021-06-04 at 8 56 36 AM" src="https://user-images.githubusercontent.com/28450084/120759022-d21aff80-c512-11eb-93dd-f37a7a79ad27.png">

<img width="575" alt="Screen Shot 2021-06-04 at 8 59 09 AM" src="https://user-images.githubusercontent.com/28450084/120759268-2625e400-c513-11eb-8548-607f66694809.png">

<img width="845" alt="Screen Shot 2021-06-04 at 9 21 38 AM" src="https://user-images.githubusercontent.com/28450084/120762182-4905c780-c516-11eb-94c4-31b21c7079ad.png">

As per the convo on Slack, I also added this new feedback screen when users create their first variation

<img width="463" alt="Screen Shot 2021-06-04 at 9 23 34 AM" src="https://user-images.githubusercontent.com/28450084/120762447-8d916300-c516-11eb-9647-1ca81e760dca.png">

Let me know if you have questions

ThomazFB committed:
Remove term list item decorator

ThomazFB committed:
Reorganize term list item UI

ThomazFB committed:
Reorganize AddAttributeTermsFragment UI

ThomazFB committed:
Extract MenuItem colorizing to dedicated extension method

anitaa1990 reviewed (approved):
LGTM :shipit: Will wait till design review is addressed before merging!

ThomazFB committed:
Define a separate layout for Attributes list on AddAttributeFragment

ThomazFB committed:
Create a base abstract Adapter for Attribute lists

ThomazFB committed:
Define AttributeAdapter for AddAttributeFragment

ThomazFB committed:
Migrate AddAttributeFragment to use new adapter structure

ThomazFB committed:
Migrate AttributeListFragment to use new adapter structure

ThomazFB committed:
Merge branch 'develop' into issue/4011-change-attributes-navigation-flow

ThomazFB commented:
Hey @Garance91540, Thanks for your feedback, I've uploaded some changes fixing your suggestions, here they are:

Fix 1
==========
![image](https://user-images.githubusercontent.com/5920403/120793257-0db2cb00-c50d-11eb-8907-16df83fd1bf5.png)

Fixed! Here's the result:
<img src="https://user-images.githubusercontent.com/5920403/120793367-2fac4d80-c50d-11eb-9df1-53e374a42615.png" width="280" height="585" />

Fix 2
==========
![image](https://user-images.githubusercontent.com/5920403/120793478-4fdc0c80-c50d-11eb-8276-5d8820742052.png)

Fixed! Here's the result:
<img src="https://user-images.githubusercontent.com/5920403/120796421-30df7980-c511-11eb-8042-0e747080e82b.png" width="280" height="585" />

PS: @Garance91540  Shouldn't the `done` button be changed to `next` only after we create the new Attributes created view? At least from the previous iteration, this view was set to have a `Done` Button

Reference used:
![image](https://user-images.githubusercontent.com/5920403/120794755-0391cc00-c50f-11eb-93e5-fae3e56bf523.png)


Fix 3
==========
![image](https://user-images.githubusercontent.com/5920403/120793840-caa52780-c50d-11eb-9bcb-346ad23f5a6f.png)

This one is getting addressed here on a dedicated PR: https://github.com/woocommerce/woocommerce-android/pull/4125

Comments
==========
I would like to suggest tackling the price warning and the new `Attributes created` view in a separate PR, since this one is becoming a large one and it's already fixing a lot of pre-existing views, would that be ok? @anitaa1990 @Garance91540

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Remove end margin from attributeListItem

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/MenuItemExt.kt:
```diff
@@ -0,0 +1,16 @@
+import android.view.MenuItem
+import androidx.core.content.ContextCompat
+
+fun MenuItem.colorizeTitle(context: Context?, colorResource: Int) = apply {
+    context?.let { ContextCompat.getColor(it, colorResource) }
+        ?.let {
+            SpannableString(this.title).apply {
+                setSpan(ForegroundColorSpan(it), 0, title.length, 0)
+            }
+        }.let { this.title = it }
+}
```
@ThomazFB I don't think we should do this for items individually, the issue @Garance91540 mentioned is impacting the whole app (check #3632).
IMO this shouldn't be fixed in this PR, nor blocking it, but should be fixed in a separate PR that fixes the whole app.

ThomazFB replied:
Thanks for this heads up @hichamboushaba, that's valuable information. I can remove this modification then and leave room for the centralized fix for the whole app. Is this ok for you @Garance91540? Can we remove this CTA color fix and leave it for an app-wide fix on another PR?

ThomazFB committed:
Remove MenuItem colorizing from Attributes CTAs

ThomazFB committed:
Improve Add Variation behavior at VariationListViewModel

ThomazFB committed:
Directly go to variation list if there's valid attributes

ThomazFB committed:
Refactor Add variation card behavior

Garance91540 reviewed (changes requested):
Hi Thomaz, 

**Fix 1**
I noticed there was a weird padding between the helper text and the "Option name" text box (see Figma for exact padding height) https://d.pr/i/Fm23fp

**Fix 2**
As "Done" signals the end of the creation flow and users [still have one more step](https://d.pr/i/5UWVW7) to carry out before generating their variation in the latest designs, "Next" here is more intuitive 

> I would like to suggest tackling the price warning and the new Attributes created view in a separate PR, since this one is becoming a large one and it's already fixing a lot of pre-existing views, would that be ok? @anitaa1990 @Garance91540

No problem!

shiki committed:
Use "Next" instead of "Done" after choosing Attributes

Next implies that there's another incoming step, which is generating the first attribute.

shiki committed:
Rename doneMenuItem to nextMenuItem

Because the title is now "Next" instead of "Done".

shiki commented:
Hi, @Garance91540! I'm taking over for Thomaz since he's AFK.

> Fix 2
As "Done" signals the end of the creation flow and users still have one more step to carry out before generating their variation in the latest designs, "Next" here is more intuitive

Done in bb718e1. Here's the before and after:

Before | After 
--------|-------
<img src="https://user-images.githubusercontent.com/198826/121054014-2cb59500-c779-11eb-9bb5-40a86b912bf6.png" width="300">       |       <img src="https://user-images.githubusercontent.com/198826/121054027-2f17ef00-c779-11eb-915d-a937da0274d1.png" width="300">

> Fix 1
I noticed there was a weird padding between the helper text and the "Option name" text box (see Figma for exact padding height) https://d.pr/i/Fm23fp

This one requires a structural change which could lead to a regression. Not to mention this PR is getting longer and we might lose some context along the way. 😅  I created a separate issue to tackle this. Is that okay? https://github.com/woocommerce/woocommerce-android/issues/4140

shiki committed:
Replace manual TextInputEditText with WCMaterialOutlinedEditTextView

shiki committed:
AddAttrTermsFrag: Replace termEditText action handling

The previous component was using EditText. We're now using WCMaterialOutline*.

shiki committed:
Revert "AddAttrTermsFrag: Replace termEditText action handling"

This reverts commit ea8c9bd0c3a4ac7efdfbdaa7e10ea5436a022320.

shiki committed:
Revert "Replace manual TextInputEditText with WCMaterialOutlinedEditTextView"

This reverts commit 87e9d991eb0938baca70fb97e96fdd4766ee98b9.

shiki commented:
@Garance91540 I created a separate PR for the padding: https://github.com/woocommerce/woocommerce-android/pull/4141

Garance91540 reviewed (approved):
Perfect, thanks! Totally ok to address the rest in a separate PR

shiki commented:
Thanks, Garance!
